src/Appwrite/Auth/OAuth2/Keycloak.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+namespace Appwrite\Auth\OAuth2;
+
+use Appwrite\Auth\OAuth2;
+
+// Reference Material
+// https://www.keycloak.org/docs/latest/securing_apps/index.html#other-openid-connect-libraries
+
+class Keycloak extends OAuth2
+{
+    /**
+     * @var array
+     */
+    protected array $scopes = [
+        'openid',
+        'profile',
+        'email',
+        'offline_access'
+    ];
+
+    /**
+     * @var array
+     */
+    protected array $user = [];
+
+    /**
+     * @var array
+     */
+    protected array $tokens = [];
+
+    /**
+     * @return string
+     */
+    public function getName(): string
+    {
+        return 'keycloak';
+    }
+
+    /**
+     * @return string
+     */
+    public function getLoginURL(): string
+    {
+        return $this->getKeycloakEndpoint() . '/realms/' . $this->getKeycloakRealm() . '/protocol/openid-connect/auth?' . \http_build_query([
+            'client_id' => $this->appID,
+            'redirect_uri' => $this->callback,
+            'state' => \json_encode($this->state),
+            'scope' => \implode(' ', $this->getScopes()),
+            'response_type' => 'code'
+        ]);
+    }
+
+    /**
+     * @param string $code
+     *
+     * @return array
+     */
+    protected function getTokens(string $code): array
+    {
+        if (empty($this->tokens)) {
+            $headers = ['Content-Type: application/x-www-form-urlencoded'];
+            $this->tokens = \json_decode($this->request(
```

Comment:
I'm trying to test this, but I'm just getting `false` back for the response from Keycloak. Any ideas what I configured incorrectly?

Reply:
I would withdraw this because of the new generell oicd provoder